### PR TITLE
90 Adding maintenance switch to Iron

### PR DESCRIPTION
**Pull request by sdaclin**

### CODE REVIEW

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java [hunk 1]
```diff
@@ -57,9 +60,9 @@
     private BigInteger m_lastSnapshotTxId = BigInteger.ONE.negate();
 
     private Disposable m_disposableTxFlow;
-    private volatile boolean m_closed = false;
 
     private final Map<String, StoreImpl> m_stores = new ConcurrentHashMap<>();
+    private State state = State.OPEN;
```
Comment (xfournet) [inline iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java:65]: should be volatile ?

Comment (sdaclin) [reply iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java:65]: Definitely !
BTW we are hiring :)

File: README.md [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [![Licence](https://img.shields.io/github/license/Axway/iron.svg)](https://www.apache.org/licenses/LICENSE-2.0.html)
-[![Build Status](https://travis-ci.org/Axway/iron.svg?branch=master)](https://travis-ci.org/Axway/iron)
+[![Build Status](https://travis-ci.com/Axway/iron.svg?branch=master)](https://travis-ci.org/Axway/iron)
```
Comment (near-ethic) [inline README.md:2]: You should update the value in parenthesis to .com too

File: iron-core/src/main/java/io/axway/iron/core/StoreManagerBuilder.java [hunk 3]
```diff
@@ -22,7 +22,7 @@ static StoreManagerBuilder newStoreManagerBuilder(String name, Properties proper
 
     StoreManagerBuilder withEntityClass(Class<?> entityClass);
 
-    <T> StoreManagerBuilder withCommandClass(Class<? extends Command<T>> commandClass);
+    <T> StoreManagerBuilder withCommandClass(Class<? extends Command<?>> commandClass);
```
Comment (near-ethic) [inline iron-core/src/main/java/io/axway/iron/core/StoreManagerBuilder.java:25]: Now that you're using ?, the <T> at the start can be removed

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerBuilderImpl.java [hunk 4]
```diff
@@ -47,7 +48,7 @@ public StoreManagerBuilder withEntityClass(Class<?> entityClass) {
     }
 
     @Override
-    public <T> StoreManagerBuilder withCommandClass(Class<? extends Command<T>> commandClass) {
+    public <T> StoreManagerBuilder withCommandClass(Class<? extends Command<?>> commandClass) {
```
Comment (near-ethic) [inline iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerBuilderImpl.java:51]: Now that you're using `<?>`, the `<T>` at the start can be removed

File: iron-inttest/src/test/java/io/axway/iron/core/FakeStoreManagerBuilderImpl.java [hunk 5]
```diff
@@ -24,7 +24,7 @@ public StoreManagerBuilder withEntityClass(Class<?> entityClass) {
     }
 
     @Override
-    public <T> StoreManagerBuilder withCommandClass(Class<? extends Command<T>> commandClass) {
+    public <T> StoreManagerBuilder withCommandClass(Class<? extends Command<?>> commandClass) {
```
Comment (near-ethic) [inline iron-inttest/src/test/java/io/axway/iron/core/FakeStoreManagerBuilderImpl.java:27]: Now that you're using `<?>`, the `<T>` at the start can be removed

File: iron-core/src/main/java/io/axway/iron/core/StoreManagerBuilder.java (hunk 3 above)
Comment (sdaclin) [reply iron-core/src/main/java/io/axway/iron/core/StoreManagerBuilder.java]: Nice

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java [hunk 6]
```diff
@@ -91,7 +94,7 @@
                 = connectableTransactions // to start only when connect is called and not with the first subscription
                 .share(); // because we subscribe the main consumer and the one in charge of notifying recovery done
 
-        Completable withTimeout = transactions.takeWhile(t -> !m_closed).timeout(1, SECONDS).ignoreElements();
+        Completable withTimeout = transactions.takeWhile(t -> !(state == State.CLOSE)).timeout(1, SECONDS).ignoreElements();
```
Comment (gvauvert) [inline iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java:97]: state != State.CLOSE ?

Comment (sdaclin) [reply iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java]: There was a rollback so we are back to !m_closed.

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sdaclin) [reply iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java]: Fixed

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java [hunk 7]
```diff
@@ -110,12 +113,30 @@
...
         m_transactionStore.close();
         m_snapshotStore.close();
     }
 
+    @Override
+    public void maintenance() {
+        LOG.info("A store manager is put in maintenance mode");
+        m_stores.forEach((storeName, store) -> {
+            try {
+                store.createCommand(MaintenanceModeCommand.class).submit().get();
```
[the remark this answers is not in the document]
Comment (sdaclin) [reply iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java]: Outdated comment

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java [hunk 8]
```diff
@@ -110,12 +113,30 @@
...
+            } catch (Exception e) {
+                throw new FormattedRuntimeException("Can't set store to maintenance mode", arguments -> arguments.add("storeName", storeName), e);
+            }
+        });
+        try {
+            snapshot();
+        } catch (Exception e) {
+            throw new FormattedRuntimeException(
+                    "Maintenance snapshot failed. Trigger a manual snapshot before restarting to recover in nominal mode (use REST API)", e);
```
[the remark this answers is not in the document]
Comment (sdaclin) [reply iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java]: Outdated

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java [hunk 9]
```diff
@@ -207,6 +234,13 @@ private void processTransaction(StorePersistence.TransactionToExecute transactio
                 try {
                     for (int i = 0; i < commands.size(); i++) {
                         Command<?> command = commands.get(i);
+                        if (command instanceof MaintenanceModeCommand) {
+                            state = State.READONLY;
+                            LOG.warn("Store enters maintenance mode (read only). Do a snapshot + restart to recover to nominal mode.");
+                        } else if (state == State.READONLY) {
+                            throw new StoreException(
+                                    "ReadWriteTransaction can't be executed, store is in maintenance mode. Do a snapshot + restart to recover to nominal mode.");
+                        }
```
[the remark this answers is not in the document]
Comment (sdaclin) [reply iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java]: Outdated

File: iron-core/src/main/java/io/axway/iron/core/spi/file/FileTransactionStore.java [hunk 10]
```diff
@@ -169,6 +173,31 @@ public void seekTransaction(BigInteger latestProcessedTransactionId) {
         initExistingFiles();
     }
 
+    @Override
+    public void lockReadOnly(boolean wantLock) {
```
[the remark this answers is not in the document]
Comment (sdaclin) [reply iron-core/src/main/java/io/axway/iron/core/spi/file/FileTransactionStore.java:177]: Should be single treaded here.

File: iron-inttest/src/test/java/io/axway/iron/spi/file/BlueGreenMigrationTest.java [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+        tryDeleteDirectory(m_txDirNodeBlue);
+        tryDeleteDirectory(m_txDirNodeGreen);
+    }
+
+    @Test
+    public final void shouldMigrateDistributedStore() throws ExecutionException, InterruptedException {
+
+        try (StoreManager storeManagerNodeBlue = initStoreManager(m_snapshotDirNodeBlue, m_txDirNodeBlue, SuperHeroV1.class, CreateSuperHeroV1.class, null)) {
+
+            Store kafkaStoreNodeBlue = storeManagerNodeBlue.getStore("kafkaStore");
```
[the remark this answers is not in the document]
Comment (sdaclin) [reply iron-inttest/src/test/java/io/axway/iron/spi/file/BlueGreenMigrationTest.java:57]: Fixed

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java [hunk 12]
```diff
@@ -162,6 +189,15 @@ public StoreImpl getStore(String storeName) {
         }
     }
 
+    private StoreImpl getSystemStore() {
+        ensureOpen();
+        try {
+            return m_stores.computeIfAbsent(SYSTEM_STORE_NAME, key -> createStore(SYSTEM_STORE_NAME));
```
[the remark this answers is not in the document]
Comment (sdaclin) [reply iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java:195]: Indeed, it's kind of an empty placeholder for now :/
Won't change for now

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java [hunk 13]
```diff
@@ -127,6 +146,7 @@ public BigInteger snapshot() {
         ensureOpen();
 
         if (m_currentTxId.compareTo(m_lastSnapshotTxId) > 0) {
+
             BigInteger tx = m_currentTxId;
 
             SnapshotPersistence snapshotPersistence = m_storePersistence.buildSnapshotPersistence(tx);
```
[the remark this answers is not in the document]
Comment (sdaclin) [reply iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java:152]: Same, won't change for now

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java [hunk 14]
```diff
@@ -246,7 +280,7 @@ private void processTransaction(StorePersistence.TransactionToExecute transactio
     }
 
     private void ensureOpen() {
-        checkState(!m_closed, "Store has been closed");
+        checkState(!(state == State.CLOSE), "Store has been closed");
```
[the remark this answers is not in the document]
Comment (sdaclin) [reply iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java]: Outdated

File: iron-inttest/src/test/java/io/axway/iron/spi/kafka/DistributedStoreMigration.java [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+            return snapshot;
+        });
+        Store kafkaStoreNodeAMigrated = storeManagerNodeAMigrated.getStore("kafkaStore");
+        Collection<SuperHeroV2> superHeroMigrated = kafkaStoreNodeAMigrated.query(readOnlyTransaction -> {
+            return readOnlyTransaction.select(SuperHeroV2.class).all();
+        });
+        assertThat(superHeroMigrated.stream().map(SuperHeroV2::nickName)).containsExactlyInAnyOrder("Hulk", "Batman");
+
+        System.out.println(superHeroMigrated);
+    }
```
[the remark this answers is not in the document]
Comment (sdaclin) [reply iron-inttest/src/test/java/io/axway/iron/spi/kafka/DistributedStoreMigration.java]: Outdated

File: iron-api/src/main/java/io/axway/iron/StoreManager.java [hunk 16]
```diff
@@ -44,10 +44,20 @@
      */
     void close();
 
+    /**
+     * Change readonly state
+     * If switching to active then a snapshot is automatically made
+     *
+     * @param active if true then switch to readonly else resume to nominal mode
+     */
+    void setReadonly(boolean active);
```
[the remark this answers is not in the document]
Comment (sdaclin) [reply iron-api/src/main/java/io/axway/iron/StoreManager.java:53]: Done

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java [hunk 17]
```diff
@@ -43,6 +46,7 @@
 
 class StoreManagerImpl implements StoreManager {
     private static final Logger LOG = LoggerFactory.getLogger(StoreManagerImpl.class);
+    public static final String SYSTEM_STORE_NAME = "iron";
```
[the remark this answers is not in the document]
Comment (sdaclin) [reply iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java:49]: let's use ironSystem then

File: iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java [hunk 18]
```diff
@@ -201,6 +237,16 @@ private void processTransaction(StorePersistence.TransactionToExecute transactio
             Object[] results = new Object[commands.size()];
             Throwable error = null;
             try {
+                if (transaction.getStoreName().equals(SYSTEM_STORE_NAME)) {
+                    Optional.ofNullable(transactionFuture).ifPresent(txFuture -> txFuture.complete(Collections.singletonList(null)));
+                    return;
+                }
+                if (transaction instanceof StorePersistence.TransactionToDiscard) {
+                    Optional.ofNullable(transactionFuture).ifPresent(txFuture -> {
+                        txFuture.completeExceptionally(new StoreException(READONLY_ERROR));
```
[the remark this answers is not in the document]
Comment (sdaclin) [reply iron-core/src/main/java/io/axway/iron/core/internal/StoreManagerImpl.java:246]: OK

File: iron-core/src/main/java/io/axway/iron/core/internal/StorePersistence.java [hunk 19]
```diff
@@ -172,6 +186,20 @@ public SnapshotPersistence buildSnapshotPersistence(BigInteger transactionId) {
...
 
+    public boolean isReadonly() {
+        return m_transactionStore.isReadOnlyLockSet();
+    }
+
+    public void setReadonly(boolean readonly) {
+        m_transactionStore.lockReadOnly(readonly);
+    }
+
+    static class TransactionToDiscard extends TransactionToExecute {
```
[the remark this answers is not in the document]
Comment (sdaclin) [reply iron-core/src/main/java/io/axway/iron/core/internal/StorePersistence.java:197]: True fact

File: iron-core/src/main/java/io/axway/iron/core/spi/file/FileTransactionStore.java [hunk 20]
```diff
@@ -169,6 +173,31 @@ public void seekTransaction(BigInteger latestProcessedTransactionId) {
...
+                Files.deleteIfExists(getReadonlyLock());
+            } catch (IOException e) {
+                throw new FormattedRuntimeException("Can't delete readonly lock", e);
+            }
+        }
+    }
+
+    @Override
+    public boolean isReadOnlyLockSet() {
+        return Files.exists(getReadonlyLock());
```
[the remark this answers is not in the document]
Comment (sdaclin) [reply iron-core/src/main/java/io/axway/iron/core/spi/file/FileTransactionStore.java:198]: File system should have a cache ;)

**Pull request merged**